File: 0384-shuffle-an-array/0384-shuffle-an-array.go
import (
    "math/rand"
)

type Solution struct {
    original []int
}


func Constructor(nums []int) Solution {
    return Solution{
        original: nums,
    }
}


func (this *Solution) Reset() []int {
    return this.original
}


func (this *Solution) Shuffle() []int {
    pool := make([]int, len(this.original))
    copy(pool, this.original)
    shuffled := make([]int, 0, len(this.original))
    for len(pool) != 0 {
        idx := rand.Intn(len(pool))
        shuffled = append(shuffled, pool[idx])
        pool = append(pool[:idx], pool[idx+1:]...)
    }
    return shuffled
}


/**
 * Your Solution object will be instantiated and called as such:
 * obj := Constructor(nums);
 * param_1 := obj.Reset();
 * param_2 := obj.Shuffle();
 */

File: 0384-shuffle-an-array/0384-shuffle-an-array.py
class Solution:

    def __init__(self, nums: List[int]):
        self.original = nums

    def reset(self) -> List[int]:
        return self.original
        

    def shuffle(self) -> List[int]:
        shuffled = []
        pool = self.original.copy()
        while pool:
            n = pool.pop(random.randrange(0, len(pool)))
            shuffled.append(n)
        return shuffled    
        
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()